### PR TITLE
Introduce platform and board hooks 

### LINKED ISSUE #75907: RFC: platform hooks to replace SYS_INIT in soc/board code
## Introduction

We currently have different way for initializing platforms, the most popular method is to call some code during the init process that initializes the platforms (soc or board) using SYS_INIT. There are a few different ways this is being done now, for example on ARM we have `CONFIG_PLATFORM_SPECIFIC_INIT` which calls a function `z_arm_platform_init` in __start before jumping to prep_c. Looking at some implementations of this function, it is not different from what is being done using SYS_INIT. Most ARM platforms just call `SystemInit();` 

All SYS_INIT calls tend to happen at the same spot during the init process and there is nothing special about this, i.e. no dependencies beside the fact that it needs to be run very early and before device and service initialization.

This RFC introduces a new way for initializing platform and moves away from SYS_INIT and arch specific implementation (on ARM) to a more generic solution using platform hooks that are always executed at the same spot. Multiple hooks will be provided to allow early and late initialization.

### Problem description

Long term, we are planning to deprecate the use of SYS_INIT and move away from priority based initialization and instead use dependency based init process. Platform initialization is fixed and does not need to be part of what is primarily used for "services". There is visible misuse of SYS_INIT here that can be replaced with something consistent and easy to implement and debug across all architectures. 
SYS_INIT is not easy to follow and debug,  priorities are arbitrary and in some cases picked randomly....

### Proposed change

Use pre-defined hooks at different stages of the boot process to allow maximum customization in a consistent way. Move all existing hooks to one unified design on all architectures.

## Detailed RFC

- provide a hook interface defined in a header, the "API" will be something that can be implemented and not called.
- identify the hook slots during the boot process and define a hook for each slot:
  - pre-c hook: something that can be called before we jump to cstart, similar to `z_arm_platform_init` right now
  - pre-kernel: very early in cstart and before we initialize device (where we have `INIT_LEVEL_EARLY` right now)
  - post -kernel: after kernel was initialized, close to where we have `INIT_LEVEL_POST_KERNEL` right now.

Likely there are other strategic spots, some boards/socs have defined interesting priorties and places to call soc/board code, however, some of these probably can be moved one of the above.

We have two options for defining such hooks:

- Use kconfig, board/soc can use kconfig and such hooks would only run if the kconfig is defined.
- Use weak symbols

We have many platforms that have such hooks (using SYS_INIT) in both soc and board code, meaning that we should consider providing the hooks for both, so instead of having one layer, for example using `platform_init_..()`, we might need to use `soc_init...()` and `board_init...()`. 

### Proposed change (Detailed)
see https://github.com/zephyrproject-rtos/zephyr/pull/62595

### Dependencies
Not many identified so fat, we need to make the change and convert all SYS_INIT call to hooks, this is currenlty WIP. We have to identify special cases and dependencies and fix those.

### Concerns and Unresolved Questions
- What hooks and where? How many hook types do we need, how flexible do we want to be.
- Kconfig vs weak symbols
- Soc/Board or generic platform hooks
- Dependency on devices and how should this be dealt with.

## Alternatives
Continue using SYS_INIT.....

### PR DESCRIPTION
**Pull request by nashif**
Introduce hook infrastructure for platform/board initialization. Replaces existing hooks implemented for few architectures and unifies them and adds hooks to replace common platform code currently implemented using SYS_INIT

resolves https://github.com/zephyrproject-rtos/zephyr/issues/75907

### CODE REVIEW

File: kernel/init.c [hunk 1]
```diff
@@ -550,6 +555,9 @@ FUNC_NORETURN void z_cstart(void)
 	/* do any necessary initialization of static devices */
 	z_device_state_init();
 
+#if CONFIG_PLATFORM_EARLY_INIT
+	platform_early_initialize();
+#endif
```
Comment (gmarull) [inline kernel/init.c:560]: wondering if this is early enough (considering EARLY exists and it is called just before entering the C domain). For example, z_device_state_init should be called after. Also, is `arch_kernel_init()` initializing something for the Kernel, or is it more like arch initial code (also required by early hooks)?

File: include/zephyr/platform/platform.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_PLATFORM_PLATFORM_H_
+#define ZEPHYR_INCLUDE_PLATFORM_PLATFORM_H_
+
+
+void platform_early_initialize(void);
+void platform_late_initialize(void);
```
Comment (gmarull) [inline include/zephyr/platform/platform.h:11]: since these are called internally, they should not be exposed in a public header. we should likely think about internal headers... now everything is under `include/zephyr`, sometimes you can see `header_internal.h`, but it's not a strict rule.

Comment (nashif) [reply include/zephyr/platform/platform.h:11]: > since these are called internally, they should not be exposed in a public header. we should likely think about internal headers... now everything is under `include/zephyr`, sometimes you can see `header_internal.h`, but it's not a strict rule.

internal to what? The call is internal, but the implementation can be external to zephyr, so we need to let implementers know what API to implement. So we need to somehow expose this as an interface (public) that can be implemented and not called somehow. We will have the same issue with other hooks, so we need to figure out how to position it and expose it.

File: kernel/init.c (hunk 1 above)
Comment (nashif) [reply kernel/init.c:560]: right now I am just trying to make this work exactly the same way as before, architecture initialization might have other caveats with multi core and SMP and other memory related activities. We probably can eliminate EARLY and make this an arch hook as well. Anyways, the positioning is subject to optimizations and change, this is more about the concept...

File: include/zephyr/platform/platform.h (hunk 2 above)
Comment (gmarull) [reply include/zephyr/platform/platform.h:11]: Only Kernel is meant to call these hooks, so placing them in `zephyr/platform.h` will make people feel they can include that file and call the functions somewhere else.

Comment (nashif) [reply include/zephyr/platform/platform.h:11]: yes, I understand, let me think about it a bit more.

File: kernel/init.c (hunk 1 above)
Comment (gmarull) [reply kernel/init.c:560]: I think the concept is ok, but we should likely merge something that doesn't change in the near term, even if it implies some slight behavior changes.

File: include/zephyr/platform/platform.h (hunk 2 above)
Comment (nashif) [reply include/zephyr/platform/platform.h:11]: some googling reveals others with similar approach :)

https://github.com/apache/nuttx/blob/master/include/nuttx/board.h

https://github.com/apache/nuttx/blob/a73f5f86dcb05ec98d66f70966f02624f2e2c9f3/include/nuttx/board.h#L103:

> * Public Function Prototypes
>  *
> * These are all standard board interfaces that are exported from board-
>  * specific logic to OS internal logic.  These should never be accessed
>  * directly from application code but may be freely used within the OS
>  *

We could have something similar,

Comment (gmarull) [reply include/zephyr/platform/platform.h:11]: Zephyr looking at NuttX :open_mouth: (just joking, I think NuttX has many well-developed concepts)

File: kernel/init.c (hunk 1 above)
Comment (dleach02) [reply kernel/init.c:560]: Why couldn't the invocation of a platform_early_initialize() just be a function run at PRE_KERNEL_1 priority 1 (or zero)?

**Comment by tbursztyka** [comment]
Missed that PR. This is a good idea, this would normalize SOC initialization across zephyr. 

Not sure we want to keep the level names in the functions however. Most of SOCs uses soc[_.]1_init() and optionally soc[_.]1_config() 
Thus what about platform_init()/platform_config() ? (or, actually, shouldn't we keep soc_init()/soc_config() instead of platform as it targets code located in soc/ ?)

**Comment by tbursztyka** [comment]
Quickly checking SYS_INIT() usage in soc/, there are 145 ones. Most of them on PRE_KERNEL_1 whether it is _init or _config 
To be more precise:
- 6 at EARLY
- 119 at PRE_KERNEL_1
- 11 at PRE_KERNEL_2
- 8 at POST_KERNEL
- 1 at APPLICATION

There are a few use case that do not land into that scheme like power initialization sometimes occurring at PRE_KERNEL_2, or some secondary SOC or "slave" (whatever this is). But these are the least present ones, they could keep using SYS_INIT().

**Comment by tbursztyka** [comment]
@nashif will you continue on this?

**Comment by nashif** [comment]
> @nashif will you continue on this?

yes

File: arch/arc/core/prep_c.c [hunk 3]
```diff
@@ -121,6 +122,10 @@ extern FUNC_NORETURN void z_cstart(void);
 
 void z_prep_c(void)
 {
+#if defined(CONFIG_PLATFORM_PREP_HOOK)
```
Comment (fabiobaltieri) [inline arch/arc/core/prep_c.c:125]: would it make sense to have weak functions instead of config options? it's already the case for other arch functions (`arch_system_halt`, `arch_spin_relax`...) would save few options, it's not like you may ever want to implement it and turn the option off anyway

Comment (henrikbrixandersen) [reply arch/arc/core/prep_c.c:125]: I think having weak defaults makes sense.

Comment (nashif) [reply arch/arc/core/prep_c.c:125]: the `__weak` approach is mentioned in the RFC as an option. We had many issues with weak functions in the past (defined at multiple levels (arch, soc, board) so you never knew which one you got etc, maybe it is not the case here. I also remember some compilers/architectures having problems with weak symbols the reason why existing hooks did not use them. Anyways, lets see if there is more feedback input to that and we can switch to weak definitions.

Comment (tbursztyka) [reply arch/arc/core/prep_c.c:125]: this way adds a lot of #ifdef, __weak would be nice yes (if there aren't any issues with it)

**Review by tbursztyka — COMMENTED** [review]
Pehaps we should take some time to get good hook names.
Should we really use short names? Why not full names (i.e. s/prep/preparation)
And trying to be more specific also: platform_boot_preparation() ?
For instance, I prefer platform_post_kernel_init() rather than platform_late_init() which is less self-documenting. It's all subjective of course, and most probably there are better proposals.

File: arch/arc/core/prep_c.c (hunk 3 above)
Comment (nashif) [reply arch/arc/core/prep_c.c:125]: at least for one of the hooks and for some platforms, the hooks is being disabled in some samples...

```
samples/drivers/ipm/ipm_mcux/remote/prj.conf:CONFIG_PLATFORM_RESET_HOOK=n
samples/subsys/ipc/openamp/remote/prj.conf:CONFIG_PLATFORM_RESET_HOOK=n
samples/subsys/ipc/openamp_rsc_table/boards/colibri_imx7d_mcimx7d_m4.conf:CONFIG_PLATFORM_RESET_HOOK=n
samples/subsys/ipc/openamp_rsc_table/boards/stm32mp157c_dk2.conf:CONFIG_PLATFORM_RESET_HOOK=n
```

need to find a out why and figure a better way to deal with this using weak symbols...

**Comment by nashif** [comment]
> Pehaps we should take some time to get good hook names.
> Should we really use short names? Why not full names (i.e. s/prep/preparation)
> And trying to be more specific also: platform_boot_preparation() ?
> For instance, I prefer platform_post_kernel_init() rather than platform_late_init() which is less self-documenting. It's all subjective of course, and most probably there are better proposals.

open for suggestions. we probably do not want to make this kernel centric, as more things happen beside kernel initialization  between the hooks.

**Review by ithinuel — DISMISSED** [review]

**Comment by ithinuel** [comment]
At first I was a bit worried by the names but the documentation in hook.h clearly states that these should not be called from the used application so I’m fine with them as is 👍.

**Comment by carlescufi** [comment]
Architecture WG:

- Proposal is welcome in general
- @henrikbrixandersen suggests changing `platform_` to `soc_` since we don't really have a clear concept of "platform" in Zephyr
- @ithinuel points to a PR with similar requirements: https://github.com/zephyrproject-rtos/zephyr/pull/71667 which uses a similar approach
- Kconfig vs weak functions: @nashif outlines pros and cons
- General consensus seems to form in favor of Kconfig instead of weak functions

**Comment by nashif** [comment]
> * @henrikbrixandersen suggests changing `platform_` to `soc_` since we don't really have a clear concept of "platform" in Zephyr

addressed

File: kernel/init.c (hunk 1 above)
Comment (tbursztyka) [reply kernel/init.c:677]: If you can hard-code the function call, you'll gain one init_entry less in ROM. (ok that's a very little amount)

SYS_INIT it meant to disappear, along with its hard-coded priority anyway.

The most important part of this PR is not where these functions are called: it could be easily fixed or rendered more flexible (whit what's going to replace SYS_INIT) since in kernel/init.c.  The API is what matters

**Review by tbursztyka — DISMISSED** [review]

File: include/zephyr/platform/hooks.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ */
+#ifndef ZEPHYR_INCLUDE_PLATFORM_PLATFORM_H_
+#define ZEPHYR_INCLUDE_PLATFORM_PLATFORM_H_
+
+/**
+ * @file
+ * @brief Platform hooks
+ *
+ * This header file contains function prototypes for the interfaces between
+ * zephyr architecture and initialization code and the platform pecific logic
```
Comment (henrikbrixandersen) [inline include/zephyr/platform/hooks.h:14]:
```suggestion
 * zephyr architecture and initialization code and the platform specific logic
```

File: include/zephyr/platform/hooks.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ * that resides under boards/ and soc/
+ *
+ * @note These are all standard platform interfaces that are exported from
+ * platform specific logic to OS internal logic.  These should never be accessed
+ * directly from application code but may be freely used within the OS.
+ */
+
+
+/**
+ * @brief Platform hook executed at the beginning of the reset vector.
```
Comment (henrikbrixandersen) [inline include/zephyr/platform/hooks.h:24]:
```suggestion
 * @brief SoC hook executed at the beginning of the reset vector.
```

Same change should be done for the remaining SoC hooks.

File: include/zephyr/platform/hooks.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * This hook is implemented by the SoC and can be used to perform any
+ * SoC-specific
```
Comment (henrikbrixandersen) [inline include/zephyr/platform/hooks.h:27]: Any SoC-specific... what? :) Same comment goes for the remaining hook definitions.

File: include/zephyr/platform/hooks.h [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * This is called before the kernel has started. This hook
+ * is implemented by the board and can be used to perform any board-specific
+ * initialization.
+ */
+void board_early_init_hook(void);
+
+/*
+ * @brief Additional platform hook executed after the kernel starts.
+
+ * This is called after the kernel has started, but before the main. This hook
```
Comment (henrikbrixandersen) [inline include/zephyr/platform/hooks.h:68]:
```suggestion
 * This is called after the kernel has started, but before the main function is called. This hook
```

File: kernel/Kconfig.init [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+menu "SoC and Board Hooks"
+
+config SOC_RESET_HOOK
+	bool "Run early SoC reset hook"
+	help
+	  Run an early SoC reset hook.
+
+	  A custom hook SoC_reset() is executed at the beginning of the
+	  startup code (__start). SoC_reset() must be implemented by the SoC.
```
Comment (henrikbrixandersen) [inline kernel/Kconfig.init:15]: `SoC_reset()`?

File: kernel/Kconfig.init [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+config SOC_PREP_HOOK
+	bool "Run early SoC preparation hook"
+	help
+	  Run an early SoC preparation hook.
+
+	  A custom hook SoC_prep() is executed at the beginning of the
+	  c prep code (prep_c). SoC_prep() must be implemented by the SoC.
```
Comment (henrikbrixandersen) [inline kernel/Kconfig.init:23]: `SoC_prep()`?

File: soc/nxp/imxrt/imxrt6xx/Kconfig [hunk 10]
```diff
@@ -6,7 +6,8 @@ config SOC_MIMXRT685S_CM33
 	select CPU_CORTEX_M33
 	select CPU_CORTEX_M_HAS_DWT
 	select CLOCK_CONTROL
-	select PLATFORM_SPECIFIC_INIT
+	select CODE_DATA_RELOCATION_SRAM if FLASH_MCUX_FLEXSPI_XIP
```
Comment (henrikbrixandersen) [inline soc/nxp/imxrt/imxrt6xx/Kconfig:9]: Was this change intended?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
A few comments regarding the provided documentation strings, but otherwise this looks like a good foundation.

File: soc/nxp/imxrt/imxrt6xx/Kconfig (hunk 10 above)
Comment (nashif) [reply soc/nxp/imxrt/imxrt6xx/Kconfig]: no, rebase mess. removed.

**Review by henrikbrixandersen — DISMISSED** [review]

File: soc/nxp/kinetis/ke1xz/Kconfig [hunk 11]
```diff
@@ -11,5 +11,4 @@ config SOC_SERIES_KE1XZ
 	select CPU_CORTEX_M_HAS_VTOR
 	select CLOCK_CONTROL
 	select HAS_MCUX
-	select PLATFORM_SPECIFIC_INIT
-	select HAS_PM
```
Comment (pillo79) [inline soc/nxp/kinetis/ke1xz/Kconfig:15]: I think this is a cut and paste error - why is HAS_PM being removed for this soc only?

Comment (nashif) [reply soc/nxp/kinetis/ke1xz/Kconfig:15]: yeah, rebase issue, fixed.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by nandojve — APPROVED** [review]
* Atmel
* GD32

**Review by decsny — APPROVED** [review]
very good, thanks for handling the NXP socs conversion

**Comment by decsny** [comment]
is there an estimated timeframe we should target to have the boards converted to this scheme by once this merges?

**Comment by nashif** [comment]
> is there an estimated timeframe we should target to have the boards converted to this scheme by once this merges?

plan is to do mass migration using a few scripts @tbursztyka  created. Some boards that need special attention will probably remain and need to be done individually.

File: kernel/Kconfig.init [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+config SOC_PREP_HOOK
+	bool "Run early SoC preparation hook"
+	help
+	  Run an early SoC preparation hook.
+
+	  A custom hook soc_prep_hook() is executed at the beginning of the
+	  c prep code (prep_c). soc_prep_hook() must be implemented by the SoC.
+
+config SOC_EARLY_INIT_HOOK
```
Comment (kartben) [inline kernel/Kconfig.init:25]: Can be a follow-up commit but would be nice if all the helps would document the name of the hook function like you did in the first two above?

**Review by bjarki-andreasen — APPROVED** [review]
Nice work!

File: arch/arm/core/Kconfig [hunk 13]
```diff
@@ -166,11 +166,15 @@ config RUNTIME_NMI
 	  needed, enable this option and attach it via z_arm_nmi_set_handler().
 
 config PLATFORM_SPECIFIC_INIT
-	bool "Platform (SOC) specific startup hook"
+	bool "Platform (SOC) specific startup hook [DEPRECATED]"
+	default y if SOC_RESET_HOOK
+	select DEPRECATED
 	help
 	  The platform specific initialization code (z_arm_platform_init) is
 	  executed at the beginning of the startup code (__start).
```
Comment (ajf58) [inline arch/arm/core/Kconfig:174]: This comment is now very misleading. `z_arm_platform_init` isn't ever called anymore.  I think this is a bug.

**Comment by ajf58** [comment]
https://github.com/zephyrproject-rtos/zephyr/issues/78386
